### PR TITLE
SQL field, table and schema matching

### PR DESCRIPTION
**Pull request by sansolovyov**
Uses new Macaw to better match Field objects used in an SQL string.

References #36911

### CODE REVIEW

File: src/metabase/native_query_analyzer.clj [hunk 1]
```diff
@@ -30,37 +29,49 @@
...
   "HoneySQL fragment to get the Field and Table"
   {:from [[:metabase_field :f]]
    ;; (t2/table-name :model/Table) doesn't work on CI since models/table.clj hasn't been loaded
    :join [[:metabase_table :t] [:= :table_id :t.id]]})
 
+(defn- field-query
+  "Exact match for quoted fields, case-insensitive match for non-quoted fields"
+  [field value]
+  (if (= (first value) \")
+    [:= field (subs value 1 (dec (count value)))]
```
Comment (tsmacdonald) [inline src/metabase/native_query_analyzer.clj:42]: it's probably a safe assumption that an opening `"` implies a closing `"` but something about this makes me a little nervous :sweat_smile:

File: src/metabase/native_query_analyzer.clj [hunk 2]
```diff
@@ -30,37 +29,49 @@
...
+    [:= [:lower field] (u/lower-case-en value)]))
+
+(defn- table-query
+  [t]
+  (if-not (:schema t)
+    (field-query :t.name (:table t))
+    [:and
+     (field-query :t.name (:table t))
+     (field-query :t.schema (:schema t))]))
```
Comment (tsmacdonald) [inline src/metabase/native_query_analyzer.clj:51]: I could see somebody complaining about the overloaded use of `field-query`, but I think it's pragmatic here, so :+1: from me

Comment (piranha) [reply src/metabase/native_query_analyzer.clj:51]: It's kind of... what can you do? We need those quotes handled and you could easily write `FROM "public".table`.

File: test/metabase/native_query_analyzer_test.clj [hunk 3]
```diff
@@ -15,3 +15,33 @@
...
+        (testing "But if you are specific - then it's a concrete field"
+          (is (= {:direct #{%venues.id} :indirect nil}
+                 (q "select v.id from venues v join checkins"))))
+        (testing "And wildcards are matching everything"
+          (is (= 10
+                 (-> (q "select * from venues v join checkins")
+                     :indirect count)))
+          (is (= 6
+                 (-> (q "select v.* from venues v join checkins")
+                     :indirect count))))))))
```
Comment (tsmacdonald) [inline test/metabase/native_query_analyzer_test.clj:47]: Compelling tests

**Review by tsmacdonald — APPROVED** [review]
Looks nice to me!

You've confirmed that our understanding of case and quotes is true for all SQL dialects we care about?

File: src/metabase/native_query_analyzer.clj (hunk 1 above)
Comment (tsmacdonald) [reply src/metabase/native_query_analyzer.clj:42]: are there crazy escape cases to worry about?

Comment (tsmacdonald) [reply src/metabase/native_query_analyzer.clj:42]: image
I think the answer is "no"

**Comment by piranha** [comment]
> You've confirmed that our understanding of case and quotes is true for all SQL dialects we care about?

Ehmm... it's a bit shameful, but I've looked at list of supported databases and just asked ChatGPT. It says that most of dialects support SQL Standard - so unquoted names are case insensitive, quoted are case sensitive.

File: src/metabase/native_query_analyzer.clj (hunk 1 above)
Comment (tsmacdonald) [reply src/metabase/native_query_analyzer.clj:42]: wait yes:

```
french_towns=# create table bar("CaseSensitive" int, CaseInSensitive text, "Perv""e""rse" int);

french_towns=# \d bar
                     Table "public.bar"
     Column      |  Type   | Collation | Nullable | Default 
-----------------+---------+-----------+----------+---------
 CaseSensitive   | integer |           |          | 
 caseinsensitive | text    |           |          | 
 Perv"e"rse      | integer |           |          | 

```

Comment (piranha) [reply src/metabase/native_query_analyzer.clj:42]: Hmmm but if we're getting `"Perv""e""erse"`, then I need to de-escape it, right? I.e. replace double-quotes with single quotes?

Comment (tsmacdonald) [reply src/metabase/native_query_analyzer.clj:42]: Oh shoot, good point

File: test/metabase/native_query_analyzer_test.clj [hunk 4]
```diff
@@ -15,3 +15,46 @@
...
+  (testing "unquoted fields are case-insensitive"
+    (is (= [:= [:lower :f.name] "test"]
+           (#'query-analyzer/field-query :f.name "test")
+           (#'query-analyzer/field-query :f.name "tEsT"))))
+  (testing "quoted fields are case-sensitive"
+    (is (= [:= :f.name "TEST"]
+           (#'query-analyzer/field-query :f.name "\"TEST\""))))
+  (testing "escaping inside quoted fields should be handled properly"
+    (is (= [:= :f.name "Perv\"e\"rse"]
+           ;; this is "Perv""e""rse"
```
Comment (tsmacdonald) [inline test/metabase/native_query_analyzer_test.clj:29]: this is indeed perverse :)

**Comment by tsmacdonald** [comment]
Oh nice, the tests are legit broken with the case-matching, I think:


```

FAIL in metabase.models.query-field-test/table-wildcard-test (query_field_test.clj:127)

with temporary :model/Card with attributes
{:creator_id 1,
 :database_id 1,
 :dataset_query {:type :native, :native {:query "SELECT NOT_TAX, TOTAL FROM orders"}, :database 1},
 :display :table,
 :name "OHFYZTENFEEYTCUGLAQY",
 :visualization_settings {},
 :created_at #t "2024-04-25T06:38:59.146814Z[UTC]",
 :updated_at #t "2024-04-25T06:38:59.146815Z[UTC]"}

 mix of select table.* and named columns
expected: 17
  actual: 16
  ```

**Comment by piranha** [comment]
> Oh nice, the tests are legit broken with the case-matching, I think:

yeah and it actually had your comment about `o.id` being wrongly guessed, so we've got our results in line. :)

**Pull request merged**